include/pstl_offload/internal/usm_memory_replacement_common.h

```diff
@@ -25,19 +26,104 @@
 namespace __pstl_offload
 {
 
+#if __linux__
+inline void*
+__original_aligned_alloc(std::size_t __alignment, std::size_t __size)
+{
+    using __aligned_alloc_func_type = void* (*)(std::size_t, std::size_t);
+
+    static __aligned_alloc_func_type __orig_aligned_alloc =
+        __aligned_alloc_func_type(dlsym(RTLD_NEXT, "aligned_alloc"));
+    return __orig_aligned_alloc(__alignment, __size);
+}
+#endif // __linux__
+
 constexpr bool
 __is_power_of_two(std::size_t __number)
 {
     return (__number != 0) && ((__number & __number - 1) == 0);
 }
 
+class __offload_policy_holder_type;
+
+// can't use std::shared_ptr, because commonly sizeof(std::shared_ptr) is 2*sizeof(void*) while
+// sizeof(__block_header) must be power of 2 and with std::shared_ptr instead of
+// __sycl_device_shared_ptr memory fragmentation increases drastically
+class __sycl_device_shared_ptr
+{
+    // state is shared between TU with the context and all objects allocated in this TU
+    struct __shared_device
+    {
+        sycl::device _M_device;
+        // to keep reference to default context of the device as long as allocated memory objects exist
+        sycl::context _M_default_context;
+        std::atomic<std::size_t> _M_cnt;
+    };
+
+    __shared_device* _M_shared_device;
+
+  public:
+    __sycl_device_shared_ptr() : _M_shared_device(nullptr) {}
+
+    explicit __sycl_device_shared_ptr(sycl::device _device)
+    {
+        _M_shared_device =
+            (__shared_device*)__original_aligned_alloc(alignof(std::max_align_t), sizeof(__shared_device));
```

Comment:
Write while need original call.

Reply:
It was a bug in `__offload_policy_holder_type` ctor: `__offload_policy_holder_mtx` was called directly while `__device_ready` was passed a ctor parameter, so they can be taken from different TUs. The result was a deadlock preventing `new` from using.